File: light_delivery/light_delivery/report/tracking_order_status/tracking_order_status.js
// Copyright (c) 2024, Muhammad Essam and contributors
// For license information, please see license.txt
 

frappe.query_reports["Tracking Order Status"] = {
    "filters": [
		{
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "reqd": 0
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 0
        },
        {
            "fieldname": "status",
            "label": __("Status"),
            "fieldtype": "Select",
            "options": [
				{ "label": __("All"), "value": "" },
				{ "label": __("Pending"), "value": "Pending" },
				{ "label": __("On The Way"), "value": "On The Way" },
                { "label": __("Delivered"), "value": "Delivered" },
                { "label": __("Cancelled"), "value": "Cancel" }
            ],
            "default": "All",
        },
		{
            "fieldname": "order_type",
            "label": __("Order Type"),
            "fieldtype": "Link",
            "options": "Order Type",
            "reqd": 0
        },
        {
            "fieldname": "store",
            "label": __("Store"),
            "fieldtype": "Link",
            "options": "Store",
            "reqd": 0
        },
        {
            "fieldname": "delivery",
            "label": __("Delivery"),
            "fieldtype": "Link",
            "options": "Delivery",
            "reqd": 0
        }
    ],

	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		// Center the text in specific columns
		if (column.fieldname === "order_id" || 
			column.fieldname === "status" || 
			column.fieldname === "order_type" || 
			column.fieldname === "delivery_name" || 
			column.fieldname === "order_date" ||
			column.fieldname === "store_name" || 
			column.fieldname === "delivery_category" ||
			column.fieldname === "total_distance" ) {
			
			value = `<div style="text-align: center;">${value}</div>`;
		}

		if (column.fieldname === "status") {
			let color = "black"; 

			if (value.includes("Pending")) {
				color = "grey";
			} else if (value.includes("Delivered")) {
				color = "green";
			} else if (value.includes("On The Way")) {
				color = "orange";
			} else if (value.includes("Delivery Cancel") || value.includes("Store Cancel") || value.includes("Cancel")) {
				color = "red";
			}

			value = `<span style="color: ${color}; font-weight: bold;">${value}</span>`;
		}

		return value;
	}

};

File: light_delivery/light_delivery/report/tracking_order_status/tracking_order_status.py
# Copyright (c) 2024, Muhammad Essam and contributors
# For license information, please see license.txt


import frappe
from frappe import _


def execute(filters=None):
    conditions = "1=1"

    if filters.get("status") == "Cancel":
        conditions += " and o.status IN ('Delivery Cancel', 'Store Cancel', 'Cancel')"
    elif filters.get("status"):
        conditions += " and o.status = %(status)s"
    if filters.get("store"):
        conditions += " and o.store = %(store)s"
    if filters.get("delivery"):
        conditions += " and o.delivery = %(delivery)s"
    if filters.get("date_range"):
        conditions += " and o.order_date between %(from_date)s and %(to_date)s"
    if filters.get("order_type"):
        conditions += " and o.order_type = %(order_type)s"


    data = frappe.db.sql("""
        select
            o.name as order_id,
            o.status,
            o.order_date,
            d.delivery_category ,
            o.phone_number as customer_mobile,
			o.order_type,
            o.delivery,
            d.delivery_name,
            o.store,
            s.store_name,
            o.total_distance
        from
            `tabOrder` o
        left join
            `tabDelivery` d on o.delivery = d.name
        left join
            `tabStore` s on o.store = s.name
        where
            {conditions}
        order by
            o.creation desc
    """.format(conditions=conditions), filters, as_dict=1)

    if data:
        for i in data:
            
            i['store_mobile'] = frappe.get_value("User",{"username":i.get("store")},'mobile_no')
            i['delivery_mobile'] = frappe.get_value("User",{"username":i.get("delivery")},'mobile_no')

    columns = [
        {"label": "Order ID", "fieldname": "order_id", "fieldtype": "Link", "options": "Order", "width": 150},
        {"label": "Delivery Category", "fieldname": "delivery_category", "fieldtype": "Link", "options": "Delivery Category", "width": 150},
        {"label": "Order Type", "fieldname": "order_type", "fieldtype": "Data", "width": 150},
		{"label": "Status", "fieldname": "status", "fieldtype": "Select", "options": "Pending\nAccepted\nDelivered\nCancelled", "width": 150},
		{"label": "Order Date", "fieldname": "order_date", "fieldtype": "Date", "width": 150},
        {"label": "Delivery Person", "fieldname": "delivery_name", "fieldtype": "Link", "options": "Delivery", "width": 150},
        {"label": "Store", "fieldname": "store_name", "fieldtype": "Link", "options": "Store", "width": 150},
        {"label": "Total Distance", "fieldname": "total_distance", "fieldtype": "Float", "width": 150},

        {"label": "Customer Mobile", "fieldname": "customer_mobile", "fieldtype": "Data", "width": 150},
        {"label": "Store Mobile", "fieldname": "store_mobile", "fieldtype": "Data", "width": 150},
        {"label": "Delivery Mobile", "fieldname": "delivery_mobile", "fieldtype": "Data", "width": 150},

    ]
    
    summary = get_report_summary(data)

    return columns, data, None, None, summary

def get_report_summary(data):
    if not data:
        return None

    total_delivered = sum(1 for row in data if row.get("status") == "Delivered")
    total_cancelled = sum(1 for row in data if row.get("status") in ["Delivery Cancel", "Store Cancel", "Cancel"])
    total_on_way = sum(1 for row in data if row.get("status") == "On The Way")
    
    return[
        {
            'value' : total_delivered,
            'indicator' : 'Green',
            'label' : _('Total: Delivered'),
            'datatype' : 'Int',
        },
        {
            'value' : total_cancelled,
            'indicator' : 'Red',
            'label' :  _('Total: Cancelled'),
            'datatype' : 'Int',
        },
        {
            'value' : total_on_way,
            'indicator' : 'Orange',
            'label' : _('Total: On the way'),
            'datatype' : 'Int',
        },
    ]